driver-core/src/test/functional/com/mongodb/client/model/expressions/ExpressionsFunctionalTest.java

```diff
@@ -0,0 +1,646 @@
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/#boolean-expression-operators
+        // (Complete as of 6.0)
+
+        // literals
+        BooleanExpression tru = Expressions.ofTrue();
+        BooleanExpression fal = Expressions.ofFalse();
+        assertExpression(true, tru, "true");
+        assertExpression(false, fal, "false");
+        assertTrue(tru instanceof BooleanExpression);
+
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/or/
+        assertExpression(true || false, tru.or(fal), "{'$or': [true, false]}");
+        assertExpression(false || true, fal.or(tru), "{'$or': [false, true]}");
+
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/and/
+        assertExpression(true && false, tru.and(fal), "{'$and': [true, false]}");
+        assertExpression(false && true, fal.and(tru), "{'$and': [false, true]}");
+
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/not/
+        assertExpression(!true, tru.not(), "{'$not': true}");
+        assertExpression(!false, fal.not(), "{'$not': false}");
+
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/cond/
+        StringExpression abc = Expressions.ofString("abc");
+        StringExpression xyz = Expressions.ofString("xyz");
+        assertExpression(
+                true && false ? "abc" : "xyz",
+                tru.and(fal).cond(abc, xyz),
+                "{'$cond': [{'$and': [true, false]}, 'abc', 'xyz']}");
+        assertExpression(
+                true || false ? "abc" : "xyz",
+                tru.or(fal).cond(abc, xyz),
+                "{'$cond': [{'$or': [true, false]}, 'abc', 'xyz']}");
+
+        // fields
+        getCollectionHelper().insertDocuments("[{ 'isTrue': true}]");
+        assertExpression(
+                true,
+                CURRENT.getFieldBoolean("isTrue"),
+                "{'$getField': {'input': '$$CURRENT', 'field': 'isTrue'}}");
+        assertExpression(
+                false,
+                CURRENT.getFieldBoolean("isTrue").not(),
+                "{'$not': {'$getField': {'input': '$$CURRENT', 'field': 'isTrue'}}}");
+        getCollectionHelper().drop();
+
+        // undefined behaviour
+        getCollectionHelper().insertDocuments("[{'isNull': null, 'isInt0': 0, 'isInt1': 1}]");
+        assertExpression(
+                true,
+                CURRENT.getFieldBoolean("isNull").not(),
+                "{'$not': {'$getField': {'input': '$$CURRENT', 'field': 'isNull'}}}");
+        assertExpression(
+                true,
+                CURRENT.getFieldBoolean("isInt0").not(),
+                "{'$not': {'$getField': {'input': '$$CURRENT', 'field': 'isInt0'}}}");
+        assertExpression(
+                false,
+                CURRENT.getFieldBoolean("isInt1").not(),
+                "{'$not': {'$getField': {'input': '$$CURRENT', 'field': 'isInt1'}}}");
```

Comment:
> null ... and undefined values

Huh, so SQL uses ternary logic, but even that was not enough for MQL, and it uses four-valued logic? Fortunately, `undefined` is deprecated, though the [docs](https://www.mongodb.com/docs/rapid/reference/mongodb-extended-json-v1/#undefined-type) explain nothing regarding the reasons behind introducing it and how it could have been useful to users.

> $not evaluates as false the following: null, ...

Given that in SQL `not null` is `null` because [`null` means "unknown"](https://www.postgresql.org/docs/current/functions-logical.html#:~:text=SQL%20uses%20a%20three%2Dvalued%20logic%20system%20with%20true%2C%20false%2C%20and%20null%2C%20which%20represents%20%E2%80%9Cunknown%E2%80%9D.), evaluating `{ $not: [ null ] }` as `true` goes against what people familiar with SQL (virtually everyone) expect.

I agree that the `$not` behavior seems as bizarre as that of the JavaScript's `==`, which is bad.

> I do not think that our API should define this behaviour

I don't see how we can forbid `CURRENT.getFieldBoolean("isInt1").not()` given that nothing here tells us they the `"isInt1"` field is not of the `Boolean` BSON type. As for not defining the behavior, I don't think we should specify the server behavior at all, because this requires us to duplicate the server docs. The only thing we should document is the correspondence between our expressions and MQL expressions with links to the corresponding server docs pages. Based on the history of discussing similar questions, I expect others to disagree with me. But please remember that each time we duplicate server docs in our docs, we introduce one more piece of docs that may eventually become outdated / inconsistent with other docs (has happened multiple times), thus potentially misleading users and undermining confidence in our docs.

Reply:
> I don't see how we can forbid `CURRENT.getFieldBoolean("isInt1").not()`

We do not forbid this, but we intentionally specify that the behaviour is undefined (which diverges from the server docs). This is intended to encourage users of our API to write better code, just as writers of modern JavaScript are told to use `===` instead of `==`.

Reply:
Considering resolved. We do not allow users to use types "as if" they were other types (or at least, we do not define behaviour in such cases, and try to prevent this).